### PR TITLE
show only opened milestones on issues page milestone filter

### LINKED ISSUE #3114: distinguish milestones on issues page (opened vs. closed)
- Gitea version (or commit ref): 311c83a
- Git version:
- Operating system: ? try.gitea.io
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the ~~bug~~ issue at https://try.gitea.io:
  - [x] Yes <https://try.gitea.io/test/milestones-issue/issues> (login is test:test)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
Sorting issues by milestones should distinguish the milestones by _opened_ and _closed_ milestones.
This would improve usability, since one rarely searches/filters for issues of already closed milestones.

A separation between _opened_ and _closed_ milestones is already done when selecting a milestone while creating an issue. I think the UI there is good and should be reused in the issues-overview page.

## Screenshots

__overview of issues, sorting by milestone:__  
milestones_sort_issues

__creating an issue, selecting milestone:__
milestones_create_issue

### PR DESCRIPTION
**Pull request by adelowo**
Fixes https://github.com/go-gitea/gitea/issues/3114

### CODE REVIEW

**Review by appleboy — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

File: models/issue_milestone.go [hunk 1]
```diff
@@ -181,7 +181,8 @@ func (milestones MilestoneList) getMilestoneIDs() []int64 {
 // GetMilestonesByRepoID returns all milestones of a repository.
 func GetMilestonesByRepoID(repoID int64) (MilestoneList, error) {
 	miles := make([]*Milestone, 0, 10)
-	return miles, x.Where("repo_id = ?", repoID).Asc("deadline_unix").Find(&miles)
+	return miles, x.Where("repo_id = ? AND is_closed = ?", repoID, 0).
```
Comment (lunny) [inline models/issue_milestone.go]: There is only one place to reference this function?

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue_milestone.go (hunk 1 above)
Comment (lunny) [reply models/issue_milestone.go]: Another reference is https://github.com/go-gitea/gitea/blob/master/routers/api/v1/repo/milestone.go#L38. The change will affect that results.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue_milestone.go (hunk 1 above)
Comment (adelowo) [reply models/issue_milestone.go]: @lunny I think It depends on expected behaviour... Should the api return both closed and opened milestones or just opened ones..

But whatever your choice, would be interested in fixing them

Comment (adelowo) [reply models/issue_milestone.go]: Or should I create ` GetOpenedMilestonesByRepoID(repoID int64)` ( names suggestion ???) so we can leave this one as it was ( keeping api backwards compatibility ).. While moving the `is_closed = 0` to the new function and use it in the issues page

Comment (lunny) [reply models/issue_milestone.go]: https://developer.github.com/v3/issues/milestones/#list-milestones-for-a-repository. It seems github's default milestones' list is opened. :( I think we should make a break change and follow github's API, both change the API docs.

Comment (adelowo) [reply models/issue_milestone.go]: Awesome then.... a break change means `kind/breaking` label ?`...  

API docs ? You mean swagger . It doesn't currently show a closed milestone in the response so we should not be needing a change ?

Comment (lunny) [reply models/issue_milestone.go]: Yes, I will add a `kind/breaking` label on this PR and I mean https://github.com/go-gitea/gitea/blob/master/routers/api/v1/repo/milestone.go#L17 and https://github.com/go-gitea/gitea/blob/master/routers/api/v1/repo/milestone.go#L21 's comments should be changed by this PR from `all of a repository's milestones` to `opened milestones of a repository`.

Comment (adelowo) [reply models/issue_milestone.go]: Done @lunny

Comment (lunny) [reply models/issue_milestone.go]: @adelowo maybe you missed comments at https://github.com/go-gitea/gitea/blob/master/routers/api/v1/repo/milestone.go#L21 and https://github.com/go-gitea/gitea/blob/master/routers/api/v1/repo/milestone.go#L17

Comment (adelowo) [reply models/issue_milestone.go]: Done @lunny ... Thanks

**Comment by adelowo** [comment]
Any idea why Drone is failing, https://drone.gitea.io/go-gitea/gitea/3215/8 ? .. This is the same error I am encountering in #5073 

`package github.com/golang/lint/golint: code in directory /go/src/github.com/golang/lint/golint expects import "golang.org/x/lint/golint"`

`make lint` works locally though

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
don't merge until v1.6 branched.

File: models/issue_milestone.go [hunk 2]
```diff
@@ -178,10 +178,11 @@ func (milestones MilestoneList) getMilestoneIDs() []int64 {
 	return ids
 }
 
-// GetMilestonesByRepoID returns all milestones of a repository.
+// GetMilestonesByRepoID returns all opened milestones of a repository.
 func GetMilestonesByRepoID(repoID int64) (MilestoneList, error) {
 	miles := make([]*Milestone, 0, 10)
-	return miles, x.Where("repo_id = ?", repoID).Asc("deadline_unix").Find(&miles)
+	return miles, x.Where("repo_id = ? AND is_closed = ?", repoID, 0).
```
Comment (lafriks) [inline models/issue_milestone.go]: it should be not `0` but `false`

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/issue_milestone.go [hunk 3]
```diff
@@ -178,10 +178,11 @@ func (milestones MilestoneList) getMilestoneIDs() []int64 {
...
+		Asc("deadline_unix").Find(&miles)
```
Comment (lafriks) [inline models/issue_milestone.go]: It should sort by deadline and then also by id as deadline is optional

File: models/issue_milestone.go (hunk 2 above)
Comment (adelowo) [reply models/issue_milestone.go]: Oh, `0` was in the db when I looked... Would fix

File: models/issue_milestone.go (hunk 3 above)
Comment (adelowo) [reply models/issue_milestone.go]: Not sure if I understand this.........

**Comment by techknowlogick** [comment]
@lafriks please review changes.

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
@lafriks Another PR should add a `state` query parameter to do that.  See https://developer.github.com/v3/issues/milestones/

**Pull request merged**